File: w5/w5d4/iteration.js
Array.prototype.bubbleSort = function() {
  var sorted = false;
  while(!sorted){
    sorted = true;
    for(var i = 0; i < this.length; i++){
      if(this[i] > this[i + 1]){
        var temp = this[i + 1];
        this[i + 1] = this[i];
        this[i] = temp;
        sorted = false;
      }
    }
  }
  return this;
};

console.log([1, 3, 2, 5, 6, 2, 10, 9].bubbleSort());

String.prototype.substrings = function() {
  var results = [];
  for(var i = 0; i < this.length; i++) {
    for(var j = i + 1; j <= this.length; j++) {
      var sub = this.slice(i, j);
      if(results.indexOf(sub) === -1){
        results.push(sub);
      }
    }
  }
  return results;
};

console.log("hello".substrings());

File: w1/w1d2/iteration.rb
def factors(number)
  lower_factors, upper_factors = [], []

  (1..Math.sqrt(number)).each do |i|
    if number % i == 0
      lower_factors << i
      upper_factors.unshift(number / i) unless (number / i) == i
    end
  end

  lower_factors + upper_factors
end

class Array
  def bubble_sort!(&prc)
    prc ||= Proc.new { |x, y| x <=> y }
    sorted = false

    until sorted
      sorted = true
      (0...self.length - 1).each do |i|
        if prc.call(self[i], self[i + 1]) == -1
          self[i], self[i + 1] = self[i + 1] , self[i]
          sorted = false
        end
      end
    end

    self
  end

  def bubble_sort(&prc)
    bubble_sort!(&prc)
  end
end

def substrings(word)
  words = []

  for i in (0...word.length)
    for j in (i...word.length)
      words << word[i..j]
    end
  end

  words
end

def subwords(word, dictionary = "dictionary.txt")
  words = File.readlines(dictionary).map(&:chomp)
  substrings(word).select { |sub| words.include?(sub) }
end
